### ISSUE TITLE
Motion sensor automation stops working, requires restarting apple TV

### ISSUE THREAD

**Issue by LP0101** [ebaauw/homebridge-hue#169]
I'm running a 4k apple TV with tvOS 11.0. Occasionally, the motion sensor-based automation will stop working, requiring me to restart my apple TV. The motion is still shown as "triggered" in the home app, though.

**Comment by ebaauw** [comment]
> requiring me to restart my apple TV

Instead of restarting the AppleTV, it should be enough to disable and re-enable HomeKit (under the iCloud account settings).

> Occasionally, the motion sensor-based automation will stop working

It's only automations triggered by _Motion Detected_ that stop working?  This looks suspiciously like the tvOS 10.2.x bug in issue #118: the AppleTV would spontaneously unregister from events for _Motion Detected_ characteristics, causing automations based on _Motion Detected_ no longer to execute.  In my (limited) testing, and from reports of others, it seems to be fixed on tvOS 11.0, at least for the 4th generation AppleTV.  I don't have a 4K AppleTV, so I cannot test that.

> Occasionally

How occasionally?  Under tvOS 10.2.x, the unregister would happen in like 5 seconds.

To confirm whether it's the same issue, could you run from the command line:
```bash
DEBUG=* homebridge -D
```
Make sure any service starting homebridge automatically is disabled or you'll get an error that the port is already in use.  In the output, you'll see debug messages from homebridge, when HomeKit clients connect to homebridge and register and unregister for notifications when characteristics change value:
```
  EventedHTTPServer [::ffff:192.168.xxx.xxx] HTTP request: /characteristics +14ms
  HAPServer [CC:22:3D:E3:CE:30] HAP Request: PUT /characteristics +0ms
  Accessory [Homebridge - mbpr] Processing characteristic set: [{"aid":47,"iid":10,"ev":true}] +0ms
  Accessory [Homebridge - mbpr] Registering Characteristic "Motion Detected" for events +0ms
  EventedHTTPServer [::ffff:192.168.xxx.xxx] HTTP Response is finished +0ms
...
  EventedHTTPServer [::ffff:192.168.xxx.xxx] HTTP request: /characteristics +10ms
  HAPServer [CC:22:3D:E3:CE:30] HAP Request: PUT /characteristics +0ms
  Accessory [Homebridge - mbpr] Processing characteristic set: [{"aid":47,"iid":10,"ev":false},{"aid":47,"iid":11,"ev":false},{"aid":47,"iid":12,"ev":false},{"aid":47,"iid":13,"ev":false},{"aid":47,"iid":14,"ev":false},{"aid":47,"iid":15,"ev":false},{"aid":47,"iid":16,"ev":false},{"aid":47,"iid":20,"ev":false},{"aid":47,"iid":21,"ev":false},{"aid":47,"iid":22,"ev":false},{"aid":47,"iid":23,"ev":false},{"aid":47,"iid":24,"ev":false},{"aid":47,"iid":25,"ev":false},{"aid":47,"iid":26,"ev":false},{"aid":47,"iid":30,"ev":false},{"aid":47,"iid":31,"ev":false},{"aid":47,"iid":32,"ev":false},{"aid":47,"iid":33,"ev":false},{"aid":47,"iid":34,"ev":false},{"aid":47,"iid":38,"ev":false},{"aid":47,"iid":39,"ev":false},{"aid":47,"iid":40,"ev":false}] +0ms
  Accessory [Homebridge - mbpr] Unregistering Characteristic "Motion Detected" for events +0ms
  Accessory [Homebridge - mbpr] Unregistering Characteristic "Last Updated" for events +0ms
```
When disabling/re-enabling HomeKit on the AppleTV, you should see similar messages that the AppleTV registers for the characteristics that you use as triggers in automations (note the IP address).  If later there is an unregister message for _Motion Detected_, this is the same bug.

As a workaround, if you still have a 4th generation AppleTV, just keep that running as additional HomeKit Hub, disconnected from any TV.  Or setup an iPad as additional HomeKit hub.

**Comment by LP0101** [comment]
Just restarting homekit didn't seem to fix it last time - a full restart of the unit seems required.

It happens haphazardly every few hours. Sometimes, it works all day without issue, though sometimes I need to restart twice to three times a day. Certainly not every 5 seconds, though.

I've started homebridge in debug mode, and will keep it running like that until it stops working again. Will paste the log here when it happens.

**Comment by LP0101** [comment]
So, over a day later, I have some updates.

First off, I'm using two home hubs now, an ipad mini 2 and the apple tv 4k. At some point, the bug happened again. This time, though, I was able to fix it just by quitting the Home app on the ipad mini and restarting it, followed by a restart of homebridge.

Attached you'll find the logfile for the past ~24 hours. I have no way of knowing exactly when the issues started, but it would be near the end. Warning, 24 mb text file.

https://transfer.sh/6BjSn/huelog

**Comment by ebaauw** [comment]
Thanks.  This is the log of `homebridge -D`, not of `DEBUG=* homebridge -D`; it doesn't contain the debug messages for homebridge communicating with HomeKit clients.

I do see some interesting stuff, though:
- You're running NodeJS v8.2.1.  In the past, I've encountered several issues with node v8 and I would really recommend using node v6.11.3 LTS.  If you need to use v8, I would upgrade to the latest v8 version;
- Homebridge-hue is polling /lights, even though no lights are exposed.  Not sure if this is a problem, you could have set `"lights": false` in config.json;
- Otherwise, everything seems normal for the first day;
- At 9/26/2017, 9:33:39 PM, homebridge-hue no longer receives responses from the Hue bridge (there's no `ok` messages for the requests and homebridge-hue doesn't issue the `get /lights` and `get /sensors` because the `get /config` failed).  It should have issued an bridge communication error message (probably a timeout), I need to check the code to see why it didn't;
- As of 9:35:17 PM, homebridge-hue again receives responses from the Hue bridge, but it appears to be corrupted.  My guess is that the `state` and `config` attributes are missing from the response to `get /sensors`, causing homebridge-hue to read all values as `null`, setting the HomeKit values accordingly.  Not just for the motion sensors, but also for the dimmer switch.  My guess is that either the Hue bridge somehow rebooted or reset itself or node v8 is playing tricks on us;
- When the sensors report their state to the Hue bridge, the `state` and `config` attributes are updated, and homebridge-hue restores the HomeKit values.  Note that there's different reports per sensor for `presence`, `lightlevel`, `temperature`, `buttonevent`, and `battery`.
- After that, everything seems normal again.

From this log, I would expect the HomeKit automation to resume working automatically after the hiccup.  It doesn't so there's more going on.  I think we can rule out the AppleTV 4K, as you're using multiple HomeKit Hubs.  To rule out node v8, could you please install node v6.11.3 LTS?  After that you need to uninstall and re-install homebridge (through npm), so it re-compiles the avahi code for node v6.

To check whether the Hue bridge reboots, you could create a CLIPGenericFlag sensor (in the [CLIP API debugger](https://developers.meethue.com/documentation/getting-started) or using [`ph.sh`](https://github.com/ebaauw/ph.sh)):
```json
{
  "manufacturername": "homebridge-hue",
  "modelid": "CLIPGenericFlag",
  "name": "Boot Time",
  "swversion": "0",
  "type": "CLIPGenericFlag",
  "uniqueid": "/sensors/xx"
}
```
and a Hue bridge rule:
```json
{
  "name": "Boot Time",
  "conditions": [
    {
      "address": "/sensors/1/state/lastupdated",
      "operator": "dx"
    },
    {
      "address": "/sensors/xx/state/flag",
      "operator": "eq",
      "value": "false"
    }
  ],
  "actions": [
    {
      "address": "/sensors/xx/state",
      "body": {
        "flag": true
      },
      "method": "PUT"
    }
  ]
}
```
This rule sets the CLIP sensor's `lastupdated` when the Hue bridge updates the `Daylight` sensor after reboot, effectively recording the boot time.  `timestriggered` should record the number of reboots.

**Comment by ebaauw** [comment]
> Homebridge-hue is polling `/lights`, even though no lights are exposed.

Fixed in v0.5.33 - also for other resource types.

**Comment by ebaauw** [comment]
Any updates, @LP0101 ?  Did you have a change to look at the error messages?
